src/Dialogs/CaldavDialog.vala

```diff
@@ -288,11 +288,35 @@ public class OnlineAccounts.CaldavDialog : Hdy.Window {
         deck.navigate (Hdy.NavigationDirection.BACK);
     }
 
-    private int sort_func (Gtk.ListBoxRow row1, Gtk.ListBoxRow row2) {
-        var cal_row1 = (CalendarRow) row1;
-        var cal_row2 = (CalendarRow) row2;
+    private void header_func (Gtk.ListBoxRow row, Gtk.ListBoxRow? before) {
+        var source_row = (SourceRow) row;
+        var is_calendar = source_row.source.has_extension (E.SOURCE_EXTENSION_CALENDAR);
+        var is_tasklist = source_row.source.has_extension (E.SOURCE_EXTENSION_TASK_LIST);
 
-        return cal_row1.source.display_name.collate (cal_row2.source.display_name);
+        Granite.HeaderLabel? header_label = null;
+
+        if (before == null) {
+            if (is_calendar) {
+                header_label = new Granite.HeaderLabel (_("Calendars"));
+            } else if (is_tasklist) {
+                header_label = new Granite.HeaderLabel (_("Task Lists"));
+            }
+
+        } else {
+            var before_source_row = (SourceRow) before;
+            var before_is_calendar = before_source_row.source.has_extension (E.SOURCE_EXTENSION_CALENDAR);
+            var before_is_tasklist = before_source_row.source.has_extension (E.SOURCE_EXTENSION_TASK_LIST);
+
+            if (before_is_calendar && is_tasklist) {
+                header_label = new Granite.HeaderLabel (_("Task Lists"));
+            } else if (before_is_tasklist && is_calendar) {
+                header_label = new Granite.HeaderLabel (_("Calendars"));
+            }
```

Comment:
Because you are adding calendars to the liststore and then adding tasklists, this seems like an impossible case